prover/src/matrix/segments.rs

```diff
@@ -0,0 +1,495 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+//
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+
+use core::iter::FusedIterator;
+use math::{fft::fft_inputs::FftInputs, FieldElement};
+use utils::collections::Vec;
+
+// CONSTANTS
+// ================================================================================================
+
+pub const ARR_SIZE: usize = 8;
+
+// SEGMENT OF ROWMAJOR MATRIX
+// ================================================================================================
+
+/// A segment of a row-major matrix of field elements. The segment is represented as a single vector
+/// of field elements, where the first element represent the first row of the segment, the element at
+/// index `i` represents the `i`-th row of the segment, and so on.
+///
+/// Each segment contains only `ARR_SIZE` columns of the matrix. For example, if we have the following
+/// matrix with 8 columns and 2 rows (the matrix is represented as a single vector of field elements
+/// in row-major order) and a ARR_SIZE of 2:
+///
+/// ```text
+/// [ 1  2  3  4  5  6  7  8 ]
+/// [ 9 10 11 12 13 14 15 16 ]
+/// ```
+/// then the first segment of this matrix is represented as a single vector of field elements:
+/// ```text
+/// [[1 2] [9 10]]
+/// ```
+/// and the second segment is represented as:
+/// ```text
+/// [[3 4] [11 12]]
+/// ```
+/// and so on.
+///
+/// It is arranged in a way that allows for efficient FFT operations.
+#[derive(Clone, Debug)]
+pub struct Segment<E>
+where
+    E: FieldElement,
+{
+    data: Vec<[E; ARR_SIZE]>,
+}
+
+#[allow(dead_code)]
+impl<E> Segment<E>
+where
+    E: FieldElement,
+{
+    // CONSTRUCTOR
+    // --------------------------------------------------------------------------------------------
+
+    /// Creates a new segment of a row-major matrix from the specified data.
+    pub fn new(data: Vec<[E; ARR_SIZE]>) -> Self {
+        Self { data }
+    }
+
+    // PUBLIC ACCESSORS
+    // --------------------------------------------------------------------------------------------
```

Comment:
Do we actually need all these accessors here? It seems like the only ones we really need are `as_data()` and `evaluate_poly()`.

Reply:
done. I've removed all the unnecessary methods from the struct implementation.